tidy3d/plugins/autograd/functions.py

```diff
@@ -575,3 +578,92 @@ def threshold(
         )
 
     return np.where(array < level, vmin, vmax)
+
+
+def smooth_max(
+    x: NDArray, tau: float = 1.0, axis: Union[int, Tuple[int, ...], None] = None
+) -> float:
+    """Compute the smooth maximum of an array using temperature parameter tau.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Input array.
+    tau : float = 1.0
+        Temperature parameter controlling smoothness. Larger values make the maximum smoother.
+    axis : Union[int, Tuple[int, ...], None] = None
+        Axis or axes over which the smooth maximum is computed. By default, the smooth maximum is computed over the entire array.
+
+    Returns
+    -------
+    np.ndarray
+        The smooth maximum of the input array.
+    """
+    return tau * logsumexp(x / tau, axis=axis)
+
+
+def smooth_min(
+    x: NDArray, tau: float = 1.0, axis: Union[int, Tuple[int, ...], None] = None
+) -> float:
+    """Compute the smooth minimum of an array using temperature parameter tau.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Input array.
+    tau : float = 1.0
+        Temperature parameter controlling smoothness. Larger values make the minimum smoother.
+    axis : Union[int, Tuple[int, ...], None] = None
+        Axis or axes over which the smooth minimum is computed. By default, the smooth minimum is computed over the entire array.
+
+    Returns
+    -------
+    np.ndarray
+        The smooth minimum of the input array.
+    """
+    return -smooth_max(-x, tau, axis=axis)
+
+
+def least_squares(
```

Comment:
i think a bit more explanation could be useful about what this returns. For example, what are the parameters being returned here? what does `func` accept (type annotation)? I assume `func` takes `params : tupl[float, ...]`? what about other args to `func` etc?

Reply:
improved the docstring and added an example